docs/standard/io/Pipelines.md

```diff
@@ -0,0 +1,802 @@
+{
+    var pipe = new Pipe();
+    Task writing = FillPipeAsync(socket, pipe.Writer);
+    Task reading = ReadPipeAsync(pipe.Reader);
+
+    return Task.WhenAll(reading, writing);
+}
+
+async Task FillPipeAsync(Socket socket, PipeWriter writer)
+{
+    const int minimumBufferSize = 512;
+
+    while (true)
+    {
+        // Allocate at least 512 bytes from the PipeWriter.
+        Memory<byte> memory = writer.GetMemory(minimumBufferSize);
+        try
+        {
+            int bytesRead = await socket.ReceiveAsync(memory, SocketFlags.None);
+            if (bytesRead == 0)
+            {
+                break;
+            }
+            // Tell the PipeWriter how much was read from the Socket.
+            writer.Advance(bytesRead);
+        }
+        catch (Exception ex)
+        {
+            LogError(ex);
+            break;
+        }
+
+        // Make the data available to the PipeReader.
+        FlushResult result = await writer.FlushAsync();
+
+        if (result.IsCompleted)
+        {
+            break;
+        }
+    }
+
+    // Tell the PipeReader that there's no more data.
+    writer.Complete();
+}
+
+async Task ReadPipeAsync(PipeReader reader)
+{
+    while (true)
+    {
+        ReadResult result = await reader.ReadAsync();
+        ReadOnlySequence<byte> buffer = result.Buffer;
+
+        while (TryReadLine(ref buffer, out ReadOnlySequence<byte> line))
+        {
+            // Process the line.
+            ProcessLine(line);
+        }
+
+        // Tell the PipeReader how much of the buffer has been consumed.
+        reader.AdvanceTo(buffer.Start, buffer.End);
+
+        // Stop reading if there's no more data.
+        if (result.IsCompleted)
+        {
+            break;
+        }
+    }
+
+    // Mark the PipeReader as complete.
+    reader.Complete();
+}
+
+bool TryReadLine(ref ReadOnlySequence<byte> buffer, out ReadOnlySequence<byte> buffer)
+{
+    // Look for a EOL in the buffer.
+    SequencePosition? position = buffer.PositionOf((byte)'\n');
+
+    if (position == null)
+    {
+        buffer = default;
+        return false;
+    }
+    
+    // Skip the line and the \n
+    buffer = buffer.Slice(buffer.GetPosition(1, position.Value));
+    return true;
+}
+```
+
+There are 2 loops:
+
+- `FillPipeAsync` reads from the `Socket` and writes to the `PipeWriter`.
+- `ReadPipeAsync` reads from the `PipeReader` and parses incoming lines.
+
+There are no explicit buffers allocated. All buffer management is delegated to the `PipeReader`/`PipeWriter` implementations. This makes it easier for consuming code to focus solely on the business logic instead of complex buffer management.
+
+In the first loop, `PipeWriter.GetMemory(int)`:
+
+* Is called to get memory from the underlying writer.
+* `PipeWriter.Advance(int)` is called to tell the `PipeWriter` how much data was written to the buffer.
+* `PipeWriter.FlushAsync` is called to make the data available to the `PipeReader`.
```

Comment:
As a reader, a question I have at this point is: why are these separate operations?

In your example, the call to `Advance` is followed immediately (as long as an exception didn't occur) by a call to `FlushAsync`. This suggests that an alternative design was possible, in which these were rolled into a single method. But presumably there's a reason you didn't do that—there must be some scenario in which you might want to call `Advance` more than once in between calls to `FlushAsync`? But what is that scenario? As it stands, I don't understand this aspect of the design.

Reply:
> This suggests that an alternative design was possible, in which these were rolled into a single method. But presumably there's a reason you didn't do that—there must be some scenario in which you might want to call Advance more than once in between calls to FlushAsync? But what is that scenario? As it stands, I don't understand this aspect of the design.

Buffering. You can GetMemory()/GetSpan and Advance(..) over and over then when you have buffered up the appropriate message/amount you call FlushAsync.

Reply:
An example https://github.com/aspnet/AspNetCore/blob/7b0b2980dc6ffb860f2bcd71868d98d85c7d35b6/src/Http/Http.Abstractions/src/Extensions/HttpResponseWritingExtensions.cs#L125-L143

Reply:
Makes sense. I don't think this is very discoverable right now, so I'm wondering if it would be helpful to add a short description of when this might be something you'd want to do. I _think_ in the app where I'm using this I don't need to, but it's hard to be certain from the docs—my read loop reads fairly large blocks of data off disk and writes them into the pipe, and I flush every time. But with the socket example you've given I don't actually know how to judge whether the "flush on every iteration" approach the code you've shown would be right in any particular scenario. If I happen to know that I'm receiving from a source that sends data in small dribs and drabs (not a hypothetical question for me, by the way) should I be waiting until I've "buffered up the appropriate message/amount"?

Reply:
It's not useful when bytes are being transferred from one source to the other. In that scenario the writer has no idea what the bytes are and can't make sense of how much should be buffered before flushing. 

There are cases like serialization that take in an `IBufferWriter<byte>` or when you're directly writing via the PipeWriter where you can choose to write an "entire message" not pieces of a message.

Here's an example: 

A simple length prefixed message, you could imagine writing code like this:

```C#
async Task WriteMessage(PipeWriter writer, byte[] payload)
{
    writer.Write(GetLength(payload));
    writer.Advance(4); // Write 4 bytes
    writer.Write(payload);
    writer.Advance(payload.Length);
    await writer.FlushAsync();
}
```